File: go/data_processing_system.go
package main

import (
	"fmt"
	"strings"
	"sync"
	"time"
)

// DataProcessingSystem orchestrates the entire data processing system
type DataProcessingSystem struct {
	taskQueue        *SharedQueue
	resultsManager   *ResultsManager
	workers          []*Worker
	numWorkerThreads int
	maxQueueSize     int
	wg               sync.WaitGroup
}

// NewDataProcessingSystem creates a new data processing system
func NewDataProcessingSystem(numWorkerThreads, maxQueueSize int, outputFileName string) *DataProcessingSystem {
	return &DataProcessingSystem{
		taskQueue:        NewSharedQueue(maxQueueSize),
		resultsManager:   NewResultsManager(outputFileName),
		workers:          make([]*Worker, 0),
		numWorkerThreads: numWorkerThreads,
		maxQueueSize:     maxQueueSize,
	}
}

// Start starts the data processing system
func (dps *DataProcessingSystem) Start() {
	fmt.Println("Starting Data Processing System...")
	fmt.Printf("Number of worker threads: %d\n", dps.numWorkerThreads)
	fmt.Printf("Maximum queue size: %d\n", dps.maxQueueSize)
	fmt.Println(strings.Repeat("=", 50))
	
	// Create and start worker threads
	for i := 1; i <= dps.numWorkerThreads; i++ {
		worker := NewWorker(fmt.Sprintf("Worker-%d", i), dps.taskQueue, dps.resultsManager)
		dps.workers = append(dps.workers, worker)
		worker.Start()
	}
	
	fmt.Println("All worker threads started successfully")
}

// AddTasks adds tasks to the processing queue
func (dps *DataProcessingSystem) AddTasks(tasks []*Task) {
	if tasks == nil || len(tasks) == 0 {
		fmt.Println("No tasks to add")
		return
	}
	
	fmt.Printf("Adding %d tasks to the queue...\n", len(tasks))
	
	for _, task := range tasks {
		added := dps.taskQueue.AddTask(task)
		if added {
			fmt.Printf("Task %d added to queue\n", task.ID)
		} else {
			fmt.Printf("Failed to add task %d to queue\n", task.ID)
		}
		
		// Small delay to simulate real-world task arrival
		time.Sleep(50 * time.Millisecond)
	}
	
	fmt.Println("Finished adding tasks to queue")
}

// Shutdown shuts down the system gracefully
func (dps *DataProcessingSystem) Shutdown() {
	fmt.Println("\nShutting down Data Processing System...")
	
	// Shutdown the task queue
	dps.taskQueue.Shutdown()
	
	// Stop all worker threads
	for _, worker := range dps.workers {
		worker.Stop()
	}
	
	// Wait for all workers to complete
	dps.wg.Add(len(dps.workers))
	for _, worker := range dps.workers {
		go func(w *Worker) {
			defer dps.wg.Done()
			w.Wait()
		}(worker)
	}
	
	// Wait for completion with timeout
	done := make(chan struct{})
	go func() {
		dps.wg.Wait()
		close(done)
	}()
	
	select {
	case <-done:
		fmt.Println("All worker threads completed")
	case <-time.After(10 * time.Second):
		fmt.Println("Timeout waiting for worker threads to complete")
	}
	
	// Write results to file
	if err := dps.resultsManager.WriteResultsToFile(); err != nil {
		fmt.Printf("Failed to write results to file: %v\n", err)
	} else {
		dps.resultsManager.PrintSummary()
	}
	
	fmt.Println("Data Processing System shutdown complete")
}

// WaitForCompletion waits for all tasks to be processed
func (dps *DataProcessingSystem) WaitForCompletion() {
	fmt.Println("Waiting for all tasks to be processed...")
	
	for !dps.taskQueue.IsEmpty() || dps.taskQueue.Size() > 0 {
		time.Sleep(100 * time.Millisecond)
	}
	
	fmt.Println("All tasks have been processed")
}

// PrintStatus prints current system status
func (dps *DataProcessingSystem) PrintStatus() {
	fmt.Println("\n=== System Status ===")
	fmt.Printf("Queue size: %d\n", dps.taskQueue.Size())
	fmt.Printf("Queue empty: %v\n", dps.taskQueue.IsEmpty())
	fmt.Printf("Queue shutdown: %v\n", dps.taskQueue.IsShutdown())
	fmt.Printf("Results count: %d\n", dps.resultsManager.GetResultCount())
	fmt.Println("===================")
}

// GetTaskQueue returns the task queue
func (dps *DataProcessingSystem) GetTaskQueue() *SharedQueue {
	return dps.taskQueue
}

// GetResultsManager returns the results manager
func (dps *DataProcessingSystem) GetResultsManager() *ResultsManager {
	return dps.resultsManager
}

File: java/DataProcessingSystem.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * Main Data Processing System that orchestrates worker threads and task processing
 */
public class DataProcessingSystem {
    private final SharedQueue taskQueue;
    private final ResultsManager resultsManager;
    private final List<WorkerThread> workers;
    private final ExecutorService executorService;
    private final int numWorkerThreads;
    private final int maxQueueSize;
    
    public DataProcessingSystem(int numWorkerThreads, int maxQueueSize, String outputFileName) {
        this.numWorkerThreads = numWorkerThreads;
        this.maxQueueSize = maxQueueSize;
        this.taskQueue = new SharedQueue(maxQueueSize);
        this.resultsManager = new ResultsManager(outputFileName);
        this.workers = new ArrayList<>();
        this.executorService = Executors.newFixedThreadPool(numWorkerThreads);
    }
    
    /**
     * Start the data processing system
     */
    public void start() {
        System.out.println("Starting Data Processing System...");
        System.out.println("Number of worker threads: " + numWorkerThreads);
        System.out.println("Maximum queue size: " + maxQueueSize);
        System.out.println("=" .repeat(50));
        
        // Create and start worker threads
        for (int i = 1; i <= numWorkerThreads; i++) {
            WorkerThread worker = new WorkerThread("Worker-" + i, taskQueue, resultsManager);
            workers.add(worker);
            executorService.submit(worker);
        }
        
        System.out.println("All worker threads started successfully");
    }
    
    /**
     * Add tasks to the processing queue
     * @param tasks List of tasks to add
     */
    public void addTasks(List<Task> tasks) {
        if (tasks == null || tasks.isEmpty()) {
            System.out.println("No tasks to add");
            return;
        }
        
        System.out.println("Adding " + tasks.size() + " tasks to the queue...");
        
        for (Task task : tasks) {
            try {
                boolean added = taskQueue.addTask(task);
                if (added) {
                    System.out.println("Task " + task.getId() + " added to queue");
                } else {
                    System.err.println("Failed to add task " + task.getId() + " to queue");
                }
                
                // Small delay to simulate real-world task arrival
                Thread.sleep(50);
                
            } catch (InterruptedException e) {
                System.err.println("Interrupted while adding tasks: " + e.getMessage());
                Thread.currentThread().interrupt();
                break;
            } catch (Exception e) {
                System.err.println("Error adding task " + task.getId() + ": " + e.getMessage());
            }
        }
        
        System.out.println("Finished adding tasks to queue");
    }
    
    /**
     * Shutdown the system gracefully
     */
    public void shutdown() {
        System.out.println("\nShutting down Data Processing System...");
        
        // Shutdown the task queue
        taskQueue.shutdown();
        
        // Stop all worker threads
        for (WorkerThread worker : workers) {
            worker.stop();
        }
        
        // Shutdown executor service
        executorService.shutdown();
        
        try {
            // Wait for all threads to complete
            if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
                System.out.println("Forcing shutdown of remaining threads...");
                executorService.shutdownNow();
                
                if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                    System.err.println("Executor service did not terminate");
                }
            }
        } catch (InterruptedException e) {
            System.err.println("Interrupted while waiting for threads to complete: " + e.getMessage());
            executorService.shutdownNow();
            Thread.currentThread().interrupt();
        }
        
        System.out.println("All worker threads completed");
        
        // Write results to file
        boolean success = resultsManager.writeResultsToFile();
        if (success) {
            resultsManager.printSummary();
        } else {
            System.err.println("Failed to write results to file");
        }
        
        System.out.println("Data Processing System shutdown complete");
    }
    
    /**
     * Wait for all tasks to be processed
     */
    public void waitForCompletion() {
        System.out.println("Waiting for all tasks to be processed...");
        
        while (!taskQueue.isEmpty() || taskQueue.size() > 0) {
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                System.err.println("Interrupted while waiting for completion: " + e.getMessage());
                Thread.currentThread().interrupt();
                break;
            }
        }
        
        System.out.println("All tasks have been processed");
    }
    
    /**
     * Get current system status
     */
    public void printStatus() {
        System.out.println("\n=== System Status ===");
        System.out.println("Queue size: " + taskQueue.size());
        System.out.println("Queue empty: " + taskQueue.isEmpty());
        System.out.println("Queue shutdown: " + taskQueue.isShutdown());
        System.out.println("Results count: " + resultsManager.getResultCount());
        System.out.println("===================");
    }
    
    /**
     * Get the task queue
     * @return The task queue
     */
    public SharedQueue getTaskQueue() {
        return taskQueue;
    }
    
    /**
     * Get the results manager
     * @return The results manager
     */
    public ResultsManager getResultsManager() {
        return resultsManager;
    }
}
